### PR TITLE
Fixed file resource raising UndefinedMethod on source_path

### LINKED ISSUE #4208: File resource source_path 'undefined method `source_path`'
```
sudo inspec exec . -t aws://
Traceback (most recent call last):
	26: from /usr/local/bin/inspec:153:in `<main>'
	25: from /usr/local/bin/inspec:153:in `load'
	24: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/inspec-bin-4.6.3/bin/inspec:11:in `<top (required)>'
	23: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
	22: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	21: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	20: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	19: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/inspec-4.6.3/lib/inspec/cli.rb:297:in `exec'
	18: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/inspec-4.6.3/lib/inspec/runner.rb:121:in `run'
	17: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/inspec-4.6.3/lib/inspec/runner.rb:114:in `load'
	16: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/inspec-4.6.3/lib/inspec/runner.rb:114:in `each'
	15: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/inspec-4.6.3/lib/inspec/runner.rb:115:in `block in load'
	14: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/inspec-4.6.3/lib/inspec/runner.rb:272:in `register_rule'
	13: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/inspec-4.6.3/lib/inspec/runner.rb:272:in `flat_map'
	12: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/inspec-4.6.3/lib/inspec/runner.rb:272:in `each'
	11: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/inspec-4.6.3/lib/inspec/runner.rb:273:in `block in register_rule'
	10: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/inspec-4.6.3/lib/inspec/runner.rb:261:in `get_check_example'
	 9: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/inspec-4.6.3/lib/inspec/runner.rb:287:in `rspec_failed_block'
	 8: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/inspec-4.6.3/lib/inspec/runner_rspec.rb:25:in `example_group'
	 7: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/rspec-core-3.8.1/lib/rspec/core/example_group.rb:260:in `block in define_example_group_method'
	 6: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/rspec-core-3.8.1/lib/rspec/core/example_group.rb:386:in `subclass'
	 5: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/rspec-core-3.8.1/lib/rspec/core/example_group.rb:422:in `set_it_up'
	 4: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/rspec-core-3.8.1/lib/rspec/core/metadata.rb:258:in `create'
	 3: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/rspec-core-3.8.1/lib/rspec/core/metadata.rb:133:in `populate'
	 2: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/rspec-core-3.8.1/lib/rspec/core/metadata.rb:180:in `build_description_from'
	 1: from /opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/inspec-4.6.3/lib/inspec/resources/file.rb:167:in `to_s'
/opt/inspec/embedded/lib/ruby/gems/2.6.0/gems/inspec-4.6.3/lib/inspec/resources/file.rb:53:in `block (2 levels) in <class:FileResource>': undefined method `source_path' for nil:NilClass (NoMethodError)
```

### PR DESCRIPTION
**Pull request by zenspider**
This is a first pass at fixing the failure found by @scottvidmar where
file resource was raising undefined_method on nil. The problem was
that the initialize method for file was not being called at all
because Resource#initialize was returning early without reporting
anything.

This fix causes a failure in test/functional/inspec_exec_test.rb where
it is testing a similar scenario but for some reason the unsupported
resources get reported. We need to figure that part out, but in the
meantime this will at least report an error at the root cause rather
than down the road.

Fixes #4208

Paired with @miah 

Signed-off-by: Ryan Davis <zenspider@chef.io>

### CODE REVIEW

**Review by jerryaldrichiii — COMMENTED** [review]
This looks great! Is this a WIP PR and we plan on adding tests for this specific use case? (e.g. a functional test of using aws target with a linux profile)?

**Comment by zenspider** [comment]
It is a WIP as far as I'm concerned because of the failing test. I don't know if the current failing test has overlap into this space or somehow they avoided each other somehow.

**Review by jerryaldrichiii — APPROVED** [review]
Approving the implementation as it exists now, pending tests.

Thanks for this. Not presenting the user with NoMethodError is valuable in its own right.

**Comment by zenspider** [comment]
I should have shown what it now does:

```
8422 % ./run.sh exec repro.rb -t aws://
Unsupported resource/backend combination: file / aws. Exiting.
```

**Comment by zenspider** [comment]
I should also say: #4211 would have saved us some serious time here because of the amount of indirectness the metaprogramming causes.

**Comment by jerryaldrichiii** [comment]
Tangentially, this may also close https://github.com/inspec/inspec/issues/4048

File: lib/inspec/plugin/v1/plugin_types/resource.rb [hunk 1]
```diff
@@ -83,11 +83,14 @@ def initialize(backend, name, *args)
           @__resource_name__ = name
 
           # check resource supports
-          supported = true
-          supported = check_supports unless @supports.nil?
+          supported = @supports ? check_supports : true
           test_backend = defined?(Train::Transports::Mock::Connection) && backend.backend.class == Train::Transports::Mock::Connection
-          # do not return if we are supported, or for tests
-          return unless supported || test_backend
+          # raise if we are supported, or for tests
```
Comment (clintoncwolfe) [inline lib/inspec/plugin/v1/plugin_types/resource.rb]: I think the comment contradicts the code, here. (I think the original did too)

**Review by clintoncwolfe — APPROVED** [review]
A minor request to fix a comment, but otherwise fine. Thanks!

**Comment by zenspider** [comment]
@clintoncwolfe I don't think this is the right fix. It does address the undefined method error bubbling up (because it was bypassing the `super` call to `initialize`) but it totally breaks ... wait.  WTF? why/how did travis pass? That should NOT be the case as far as I'm concerned. Here is what I see if I run this  file by itself or this one test by itself:

```
  1) Failure:
inspec exec::with a profile that contains skipped resources#test_0001_exits with an error [/Users/ryan/Work/git/inspec/inspec/test/functional/inspec_exec_test.rb:221]:
Expected "[2019-06-17T19:49:21-07:00] WARN: DEPRECATION: AWS resources shipped with core InSpec are being to moved to a resource pack for faster iteration. Please update your profiles to depend on git@github.com:inspec/inspec-aws.git . Resource 'aws_iam_users' (used at /Users/ryan/Work/git/inspec/inspec/test/unit/mock/profiles/aws-profile/controls/aws.rb:3)
[2019-06-17T19:49:21-07:00] WARN: DEPRECATION: AWS resources shipped with core InSpec are being to moved to a resource pack for faster iteration. Please update your profiles to depend on git@github.com:inspec/inspec-aws.git . Resource 'aws_iam_access_keys' (used at /Users/ryan/Work/git/inspec/inspec/test/unit/mock/profiles/aws-profile/controls/aws.rb:10)
[2019-06-17T19:49:21-07:00] WARN: DEPRECATION: AWS resources shipped with core InSpec are being to moved to a resource pack for faster iteration. Please update your profiles to depend on git@github.com:inspec/inspec-aws.git . Resource 'aws_s3_bucket' (used at /Users/ryan/Work/git/inspec/inspec/test/unit/mock/profiles/aws-profile/controls/aws.rb:17)

Profile: aws example profile (aws)
Version: 1.0.0
Target:  local://

\e[38;5;9m  ×  Users that have a password but do not have MFA enabled: /Users/ryan/Work/git/inspec/inspec/test/unit/mock/profiles/aws-profile/controls/aws.rb:1\e[0m
\e[38;5;9m     ×  Control Source Code Error /Users/ryan/Work/git/inspec/inspec/test/unit/mock/profiles/aws-profile/controls/aws.rb:1 
     undefined method `aws_client' for #<Train::Transports::Local::Connection:0x00007fdaee127200>\e[0m
\e[38;5;9m  ×  Do not allow access keys older than 90 days: /Users/ryan/Work/git/inspec/inspec/test/unit/mock/profiles/aws-profile/controls/aws.rb:8\e[0m
\e[38;5;9m     ×  Control Source Code Error /Users/ryan/Work/git/inspec/inspec/test/unit/mock/profiles/aws-profile/controls/aws.rb:8 
     undefined method `aws_client' for #<Train::Transports::Local::Connection:0x00007fdaee127200>\e[0m
\e[38;5;9m  ×  Check our demo bucket for dangerous settings: /Users/ryan/Work/git/inspec/inspec/test/unit/mock/profiles/aws-profile/controls/aws.rb:15\e[0m
\e[38;5;9m     ×  Control Source Code Error /Users/ryan/Work/git/inspec/inspec/test/unit/mock/profiles/aws-profile/controls/aws.rb:15 
     undefined method `aws_client' for #<Train::Transports::Local::Connection:0x00007fdaee127200>\e[0m


Profile Summary: 0 successful controls, \e[38;5;9m3 control failures\e[0m, 0 controls skipped
Test Summary: 0 successful, \e[38;5;9m3 failures\e[0m, 0 skipped
" to include "Resource `aws_iam_users` is not supported on platform".
```

(converted newlines for readability)

Basically, my `ArgumentError` prevents those other exceptions from being aggregated as test results.

**Comment by zenspider** [comment]
Here's what I'm doing:

```
m test/functional/inspec_exec_test.rb:221 | ruby -pe 'gsub(/\\n/, "\n")'
```

File: lib/inspec/plugin/v1/plugin_types/resource.rb [hunk 2]
```diff
@@ -83,11 +83,14 @@ def initialize(backend, name, *args)
...
+          unless supported || test_backend
+            msg = "Unsupported resource/backend combination: %s / %s. Exiting." %
+              [name, backend.platform.name]
+            raise ArgumentError, msg
```
Comment (clintoncwolfe) [inline lib/inspec/plugin/v1/plugin_types/resource.rb:97]: Here's an idea, which would at least improve messaging - create a custom exception class (or we may have a suitable one already - see lib/exceptions or lib/errors, I know we have two). Perhaps ResourceUnsupportedError? Then add a `rescue` matching that new exception class in the control evaluator - https://github.com/inspec/inspec/blob/954e6c0ff98113ed41e95b50f5a6651a30572c8c/lib/inspec/rule.rb#L54 . It make a special case of this, but would give you the chance to handle the outcome with a little more finesse.

**Review by miah — CHANGES_REQUESTED** [review]
Lets pair on this, I think we should try to bubble this up through another exception. If we can't this is fine for now, but we should file a issue to revisit this exit path in the future.

**Comment by zenspider** [comment]
happy to later today
